### PR TITLE
Add some snow

### PR DESCRIPTION
**Pull request by hovertank3d**

### CODE REVIEW

File: src/crispy.h [hunk 1]
```diff
@@ -64,8 +64,9 @@ typedef struct
 	int overunder;
 	int pitch;
 	int playercoords;
-	int recoil;
-	int secretmessage;
+    int recoil;
```
Comment (fabiangreffrath) [inline src/crispy.h:67]: Please keep indentation intact.

File: src/doom/d_main.c [hunk 2]
```diff
@@ -79,6 +80,10 @@
 
 #include "d_main.h"
 
+typedef struct snowflake_t {
+    int x, y;
+}snowflake_t;
```
Comment (fabiangreffrath) [inline src/doom/d_main.c:85]: Space after the bracelet, please.

File: src/doom/d_main.c [hunk 3]
```diff
@@ -167,6 +172,8 @@ void D_ProcessEvents (void)
 
 // wipegamestate can be set to -1 to force a wipe on the next draw
 gamestate_t     wipegamestate = GS_DEMOSCREEN;
+snowflake_t     snowflakes[5000];
```
Comment (fabiangreffrath) [inline src/doom/d_main.c:175]: Could you make this a function of SCREENHEIGHT and SCREENWIDTH, please? Something like `(SCREENHEIGHT * SCREENWIDTH) >> 6`. This will probably require dynamic allocation.

File: src/doom/d_main.c [hunk 4]
```diff
@@ -182,9 +189,18 @@ boolean D_Display (void)
     int				y;
     boolean			wipe;
     boolean			redrawsbar;
-		
+
```
Comment (fabiangreffrath) [inline src/doom/d_main.c:192]: Please, no unrelated whitespace changes here and in the rest of this file.

File: src/doom/d_main.c [hunk 5]
```diff
@@ -182,9 +189,18 @@ boolean D_Display (void)
...
+    if(!start_snoflakes && crispy->snowflakes)
+    {
+        for(size_t i = 0; i < 5000; i++)
```
Comment (fabiangreffrath) [inline src/doom/d_main.c:195]: Does this survive zn in-game High Resolution or Widescreen Rendering switch?

File: src/doom/d_main.c [hunk 6]
```diff
@@ -296,12 +312,33 @@ boolean D_Display (void)
     inhelpscreensstate = inhelpscreens;
     oldgamestate = wipegamestate = gamestate;
     
+
```
Comment (fabiangreffrath) [inline src/doom/d_main.c:315]: Excess blank line.

File: src/doom/d_main.c [hunk 7]
```diff
@@ -296,12 +312,33 @@ boolean D_Display (void)
...
+    if(crispy->snowflakes)
+    {
```
Comment (fabiangreffrath) [inline src/doom/d_main.c:317]: No mixed declaration and code, please.

File: src/doom/m_menu.c [hunk 8]
```diff
@@ -603,6 +603,10 @@ enum
     crispness_demousetimer,
     crispness_sep_demos_,
 
+    crispness_sep_new_year,
```
Comment (fabiangreffrath) [inline src/doom/m_menu.c:606]: No chance this gets a switch in the menu. Sorry, but this feature has to remain secret. 😉

**Review by fabiangreffrath — CHANGES_REQUESTED** [review]
This must be a joke. But I like it. 😄 

Please get rid of all unrelated whitespace changes znd address the issues I pointed out.

File: src/doom/d_main.c [hunk 9]
```diff
@@ -296,12 +312,33 @@ boolean D_Display (void)
...
+            snowflakes[i].x += 1 - Crispy_Random() % 3;
+            if(snowflakes[i].y >= SCREENHEIGHT)
+                snowflakes[i].y = 0;
+            if(snowflakes[i].x >= SCREENWIDTH)
+                snowflakes[i].x = 0;
+            if(snowflakes[i].x < 0)
+                snowflakes[i].x = SCREENWIDTH;
+
+            I_VideoBuffer[snowflakes[i].x + snowflakes[i].y * SCREENWIDTH] = 4;
+        }
```
Comment (fabiangreffrath) [inline src/doom/d_main.c:333]: Better determine the best matching snow flake color once during initialization at run-time. Some PWADs change the palette in unexpected ways.

File: src/doom/d_main.c (hunk 3 above)
Comment (hovertank3d) [reply src/doom/d_main.c:175]: Ok!

File: src/crispy.h (hunk 1 above)
Comment (hovertank3d) [reply src/crispy.h:67]: I don't know why whitespaces changed. Maybe IDE did it...
I'll fix it nowю

File: src/doom/d_main.c (hunk 5 above)
Comment (hovertank3d) [reply src/doom/d_main.c:195]: Yes 🤔 
it works nice for me

File: src/doom/m_menu.c (hunk 8 above)
Comment (hovertank3d) [reply src/doom/m_menu.c:606]: It will New Year egg (:

**Comment by hovertank3d** [comment]
I have fixed code. Can you check it?

**Comment by hovertank3d** [comment]
WTH
i need to learn git

File: src/crispy.h [hunk 10]
```diff
@@ -66,6 +66,7 @@ typedef struct
 	int playercoords;
 	int recoil;
 	int secretmessage;
+	int snowflakes;
```
Comment (fabiangreffrath) [inline src/crispy.h:69]: If this isn't a config key anymore, it should get moved down into the "in-game switches and variables" section.

File: src/doom/d_main.c [hunk 11]
```diff
@@ -237,20 +266,43 @@ boolean D_Display (void)
...
         // [crispy] Crispy HUD
-        if (screenblocks >= CRISPY_HUD)
+        if (screenblocks >= CRISPY_HUD || crispy->snowflakes)
             ST_Drawer(false, true);
     }
 
+    // [crispy] draw snowflakes
+    if(crispy->snowflakes)
+    {
+	int color = I_GetPaletteIndex(0xFF, 0xFF, 0xFF);
```
Comment (fabiangreffrath) [inline src/doom/d_main.c:286]: Only once, please. Declare it as `static`, initialize with `-1` and only recalculate if it's `-1`.

File: src/doom/d_main.c [hunk 12]
```diff
@@ -237,20 +266,43 @@ boolean D_Display (void)
 	D_PageDrawer ();
 	break;
     }
-
```
Comment (fabiangreffrath) [inline src/doom/d_main.c:240]: Unrelated whitespace changes.

File: src/doom/d_main.c [hunk 13]
```diff
@@ -295,7 +347,6 @@ boolean D_Display (void)
     viewactivestate = viewactive;
     inhelpscreensstate = inhelpscreens;
     oldgamestate = wipegamestate = gamestate;
-
```
Comment (fabiangreffrath) [inline src/doom/d_main.c:298]: Unrelated whitespace changes.

File: src/doom/d_main.c [hunk 14]
```diff
@@ -459,6 +509,7 @@ void D_BindVariables(void)
     M_BindIntVariable("crispy_pitch",           &crispy->pitch);
     M_BindIntVariable("crispy_playercoords",    &crispy->playercoords);
     M_BindIntVariable("crispy_secretmessage",   &crispy->secretmessage);
+    M_BindIntVariable("crispy_snowflakes",      &crispy->snowflakes);
```
Comment (fabiangreffrath) [inline src/doom/d_main.c:512]: This isn't a config variable anymore.

File: src/doom/d_main.c [hunk 15]
```diff
@@ -2027,6 +2078,8 @@ void D_DoomMain (void)
     I_InitSound(true);
     I_InitMusic();
 
+	crispy->snowflakes = 0; // [crispy] turn off by default
```
Comment (fabiangreffrath) [inline src/doom/d_main.c:2081]: This is not necessary, as the whole `crispy` struct gets initialized.

File: src/doom/m_crispy.h [hunk 16]
```diff
@@ -71,6 +71,7 @@ extern void M_CrispyToggleOverunder(int choice);
 extern void M_CrispyTogglePitch(int choice);
 extern void M_CrispyTogglePlayerCoords(int choice);
 extern void M_CrispyToggleSecretmessage(int choice);
+extern void M_CrispyToggleSnow(int choice);
```
Comment (fabiangreffrath) [inline src/doom/m_crispy.h:74]: Not necessary anymore.

File: src/doom/st_stuff.c [hunk 17]
```diff
@@ -451,6 +451,7 @@ cheatseq_t cheat_showfps2 = CHEAT("idrate", 0); // [crispy] PrBoom+
 cheatseq_t cheat_goobers = CHEAT("goobers", 0);
 cheatseq_t cheat_version = CHEAT("version", 0); // [crispy] Russian Doom
 cheatseq_t cheat_skill = CHEAT("skill", 0);
+cheatseq_t cheat_snow = CHEAT("snow", 0);
```
Comment (fabiangreffrath) [inline src/doom/st_stuff.c:454]: I like the idea of turning it in to a cheat code. However, "snow" is a but generic, isn't it? Put some music into it, how about "letitsnow"? :smile:

File: src/m_config.c [hunk 18]
```diff
@@ -2179,6 +2179,14 @@ static default_t extra_defaults_list[] =
 
     CONFIG_VARIABLE_INT(crispy_secretmessage),
 
+    //!
+    // @game doom
+    //
+    // Snow!.
+    //
+
+    CONFIG_VARIABLE_INT(crispy_snowflakes),
```
Comment (fabiangreffrath) [inline src/m_config.c:2188]: It isn't a config variable anymore.

**Review by fabiangreffrath — CHANGES_REQUESTED** [review]
Thank you for your improvements!

However, there is one issue left and this one is crucial: You update the snow flakes' coordinates in `D_Display()` which means the speed of the snowflakes falling depends on the rendering framerate.This may be up to several hundred FPS is Uncapped Framerate is enabled and VSync disabled - as opposed to 35 FPS in regular mode.

What's needed is to split this up into two functions, one for the actual drawing of the snow flakes and one for updating their coordinates. That second one must be called from a place that's only processed once per tic, e.g. `P_UpdateSpecials()` or similar.

This really needs to be addressed before anything can get merged. And with this raised level of complexity, maybe it even makes sense to split the whole snowflake processing into its own source file?

**Comment by hovertank3d** [comment]
> And with this raised level of complexity, maybe it even makes sense to split the whole snowflake processing into its own source file?

So, can this files be named something like d_snow.{c,h}?

File: src/doom/CMakeLists.txt [hunk 19]
```diff
@@ -78,6 +78,7 @@ add_library(doom STATIC
             statdump.c      statdump.h
             st_lib.c        st_lib.h
             st_stuff.c      st_stuff.h
+	    v_snow.c        v_snow.h
```
Comment (fabiangreffrath) [inline src/doom/CMakeLists.txt:81]: Please fix indentation.

File: src/doom/d_main.c [hunk 20]
```diff
@@ -182,9 +184,9 @@ boolean D_Display (void)
     int				y;
     boolean			wipe;
     boolean			redrawsbar;
-		
-    redrawsbar = false;
     
+    redrawsbar = false;
+
```
Comment (fabiangreffrath) [inline src/doom/d_main.c:189]: Here is still some unrelated whitespace change.

File: src/doom/d_main.c [hunk 21]
```diff
@@ -325,7 +331,6 @@ boolean D_Display (void)
                           W_CacheLumpName (DEH_String("M_PAUSE"), PU_CACHE));
     }
 
-
```
Comment (fabiangreffrath) [inline src/doom/d_main.c:328]: Here is another unrelated whitespace change.

File: src/doom/d_main.c [hunk 22]
```diff
@@ -246,8 +248,12 @@ boolean D_Display (void)
     {
 	R_RenderPlayerView (&players[displayplayer]);
 
+	// [crispy] Snow
+	if(crispy->snowflakes)
+	    V_DrawSnow();
+
         // [crispy] Crispy HUD
-        if (screenblocks >= CRISPY_HUD)
+        if (screenblocks >= CRISPY_HUD || crispy->snowflakes)
```
Comment (fabiangreffrath) [inline src/doom/d_main.c:256]: So, does this redraw the whole status bar background or only the status bar numbers?

File: src/doom/v_snow.c [hunk 23]
```diff
@@ -0,0 +1,121 @@
+//
+// Copyright(C) 1993-1996 Id Software, Inc.
+// Copyright(C) 2005-2014 Simon Howard
+// Copyright(C) 2015-2018 Fabian Greffrath
```
Comment (fabiangreffrath) [inline src/doom/v_snow.c:4]: 2021, please. Where is your own copyright?

File: src/doom/v_snow.c [hunk 24]
```diff
@@ -0,0 +1,121 @@
...
+    int x, y;
+} snowflake_t;
+
+static snowflake_t     *snowflakes = NULL;
+static size_t          snowflakes_num;
+static int             snowflakes_color = -1;
+static int             last_screen_size;
+static int             wind = 1;
+
+static bool CoordsNeedsUpdate()
```
Comment (fabiangreffrath) [inline src/doom/v_snow.c:39]: This function duplicates all the code that is needed to check if a gametic has passed. Please do simply update the snow coords in `P_UpdateSpecials()` as suggested earlier.

File: src/doom/v_snow.c [hunk 25]
```diff
@@ -0,0 +1,121 @@
...
+	snowflakes[i].x = rand() % (SCREENWIDTH);
+    }
+}
+
+static void UpdateSnowCoords()
+{
+    if (rand()%150 == 1)
+	wind = 1 - rand()%3;
+
+    for (size_t i = 0; i < snowflakes_num; i++)
```
Comment (fabiangreffrath) [inline src/doom/v_snow.c:69]: Indentation is somehow messed up in this file. Could you please just tab-indent all the new code.

File: src/doom/v_snow.c [hunk 26]
```diff
@@ -0,0 +1,121 @@
...
+    
+    last_screen_size = SCREENWIDTH * SCREENHEIGHT;
+    snowflakes_num = last_screen_size >> 6;
+    snowflakes = Z_Malloc(snowflakes_num * sizeof(snowflake_t), PU_STATIC, NULL);
+
+    InitSnowCoords();
+
+    if (snowflakes_color == -1)
+	snowflakes_color = I_GetPaletteIndex(0xFF, 0xFF, 0xFF);
+
```
Comment (fabiangreffrath) [inline src/doom/v_snow.c:100]: Here is an excess blank line.

File: src/doom/v_snow.h [hunk 27]
```diff
@@ -0,0 +1,25 @@
+//
+// Copyright(C) 1993-1996 Id Software, Inc.
+// Copyright(C) 2005-2014 Simon Howard
+// Copyright(C) 2015-2018 Fabian Greffrath
```
Comment (fabiangreffrath) [inline src/doom/v_snow.h:4]: Please update my Copyright to 2021 and add your own here as well.

**Review by fabiangreffrath — CHANGES_REQUESTED** [review]
Vast improvement! Sorry to stress your patience even further, but I still have to request some changes,

File: src/doom/d_main.c (hunk 22 above)
Comment (hovertank3d) [reply src/doom/d_main.c:256]: Yes, it redrawing whole status bar

File: src/doom/v_snow.c (hunk 24 above)
Comment (hovertank3d) [reply src/doom/v_snow.c:39]: Oh I didn't notice it 😄

File: src/doom/d_main.c [hunk 28]
```diff
@@ -52,6 +52,7 @@
 #include "m_controls.h"
 #include "m_misc.h"
 #include "m_menu.h"
+#include "m_random.h"
```
Comment (fabiangreffrath) [inline src/doom/d_main.c:55]: This isn't necessary anymore.

File: src/doom/v_snow.c [hunk 29]
```diff
@@ -0,0 +1,102 @@
...
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// DESCRIPTION:
+//	[crispy] Snow drawing
+//
+
+#include <stdlib.h>
+#include <stdbool.h>
```
Comment (fabiangreffrath) [inline src/doom/v_snow.c:22]: Why stdbool? If you need a bool type (do you?) please include doomtype.h.

File: src/doom/v_snow.c [hunk 30]
```diff
@@ -0,0 +1,102 @@
...
+    {
+        snowflakes[i].y = 0 - rand() % (SCREENHEIGHT * 2);
+        snowflakes[i].x = rand() % (SCREENWIDTH);
+    }
+}
+
+void V_SnowUpdate()
+{
+    if (rand() % 150 == 1)
+        wind = 1 - rand() % 3;
```
Comment (fabiangreffrath) [inline src/doom/v_snow.c:52]: Please use M_CrispyRandom(). It's meant for exactly this kind of stuff.

File: src/doom/v_snow.c [hunk 31]
```diff
@@ -0,0 +1,102 @@
...
+            snowflakes[i].x = snowflakes[i].x - SCREENWIDTH;
+        if (snowflakes[i].x < 0)
+            snowflakes[i].x = SCREENWIDTH - snowflakes[i].x;
+    }
+}
+
+static void ResetSnow()
+{
+    if (snowflakes != NULL)
+        Z_Free(snowflakes);
```
Comment (fabiangreffrath) [inline src/doom/v_snow.c:75]: Please use stdlib's malloc() and realloc() instead. I'd like to avoid stressing Doom's already rather limited zone memory even more.

**Review by fabiangreffrath — CHANGES_REQUESTED** [review]

File: src/doom/v_snow.c [hunk 32]
```diff
@@ -0,0 +1,102 @@
...
+
+    if (snowflakes_color == -1)
+        snowflakes_color = I_GetPaletteIndex(0xFF, 0xFF, 0xFF);
+}
+
+void V_SnowDraw()
+{
+    int video_offset;
+
+    if (last_screen_size != (SCREENHEIGHT * SCREENWIDTH))
```
Comment (fabiangreffrath) [inline src/doom/v_snow.c:91]: I think this can go into V_SnowUpdate().

File: src/doom/d_main.c [hunk 33]
```diff
@@ -246,8 +247,12 @@ boolean D_Display (void)
     {
 	R_RenderPlayerView (&players[displayplayer]);
 
+	// [crispy] Snow
+	if(crispy->snowflakes)
```
Comment (fabiangreffrath) [inline src/doom/d_main.c:251]: `if (` please

File: src/doom/v_snow.c [hunk 34]
```diff
@@ -0,0 +1,105 @@
...
+static int wind = 1;
+
+static void InitSnowCoords()
+{
+    for (size_t i = 0; i < snowflakes_num; i++)
+    {
+        // here rand() but not Crispy_Random() because
+        // Crispy_Random() returns too small numbers
+        snowflakes[i].y = 0 - rand() % (SCREENHEIGHT * 2);
+        snowflakes[i].x = rand() % (SCREENWIDTH);
```
Comment (fabiangreffrath) [inline src/doom/v_snow.c:47]: How about `M_CrispyRandom() * SCREENWIDTH / 256` and likewise above.

File: src/doom/v_snow.c [hunk 35]
```diff
@@ -0,0 +1,105 @@
...
+    }
+}
+
+static void ResetSnow()
+{
+
```
Comment (fabiangreffrath) [inline src/doom/v_snow.c:53]: Blank line.

File: src/doom/v_snow.c [hunk 36]
```diff
@@ -0,0 +1,105 @@
...
+    last_screen_size = SCREENWIDTH * SCREENHEIGHT;
+    snowflakes_num = last_screen_size >> 6;
+
+    if (snowflakes != NULL)
+        free(snowflakes);
+    
+    snowflakes = malloc(snowflakes_num * sizeof(snowflake_t));
```
Comment (fabiangreffrath) [inline src/doom/v_snow.c:60]: No need to `free()` and `alloc()`. `realloc(NULL, ...)` is the same as `alloc(...)`.

File: src/doom/v_snow.h [hunk 37]
```diff
@@ -0,0 +1,26 @@
...
+// DESCRIPTION:
+//	[crispy] Snow drawing
+//
+
+#ifndef __V_SNOW__
+#define __V_SNOW__
+
+void V_SnowDraw();
+void V_SnowUpdate();
+#endif
```
Comment (fabiangreffrath) [inline src/doom/v_snow.h:26]: Could you add one blank line before the `#endif`?

File: src/doom/st_stuff.c [hunk 38]
```diff
@@ -1161,6 +1162,11 @@ ST_Responder (event_t* ev)
 	    plyr->message = msg;
 	}
       }
+	// [crispy] snow
```
Comment (fabiangreffrath) [inline src/doom/st_stuff.c:1165]: This doesn't seem to be indented as the paragraph above.

File: src/doom/v_snow.c [hunk 39]
```diff
@@ -0,0 +1,105 @@
...
+
+        if (snowflakes[i].y >= SCREENHEIGHT)
+            snowflakes[i].y = 0;
+        if (snowflakes[i].x >= SCREENWIDTH)
+            snowflakes[i].x = snowflakes[i].x - SCREENWIDTH;
+        if (snowflakes[i].x < 0)
+            snowflakes[i].x = SCREENWIDTH + snowflakes[i].x;
+    }
+}
+void V_SnowDraw()
```
Comment (fabiangreffrath) [inline src/doom/v_snow.c:93]: Please add one blank line between functions.

**Review by fabiangreffrath — CHANGES_REQUESTED** [review]
Finishing straight 😉

File: src/doom/v_snow.c (hunk 36 above)
Comment (hovertank3d) [reply src/doom/v_snow.c:60]: I think it's better to use free() cause snow animation should to be restarted when the resolution is changed.

**Comment by fabiangreffrath** [comment]
Thank you for your patience!

**Comment by hovertank3d** [comment]
Happy Holidays!
